KAFKA-3144: Report members with no assigned partitions in ConsumerGroupCommand

vahidhashemian opened the pull request:
This PR makes a couple of enhancements to the `--describe` option of `ConsumerGroupCommand`:
1. Listing members with no assigned partitions.
2. Showing the member id along with the owner of each partition (owner is supposed to be the logical application id and all members in the same group are supposed to set the same owner).
3. Printing a warning indicating whether ZooKeeper based or new consumer API based information is being reported.

It also adds unit tests to verify the added functionality.

Note: The third request on the corresponding JIRA (listing active offsets for empty groups of new consumers) is not implemented as part of this PR, and has been moved to its own JIRA (KAFKA-3853).

vahidhashemian commented:
@junrao Would appreciate your input on this pull request.

junrao commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -156,47 +163,102 @@ object ConsumerGroupCommand {
+      // mapping of topic partition -> ephemeral owner
+      val eOwnerByTopicPartition = scala.collection.mutable.Map[TopicAndPartition, Long]()
+      // mapping of topic partition -> owner
+      val ownerByTopicPartition = scala.collection.mutable.Map[TopicAndPartition, String]()
```
Reporting both owner and member-id can be a bit confusing. Also, for ZK based consumer, we get the following output. The member-id part is repeated in the owner part. Perhaps instead, we can report 3 fields: member-id, client-id, and client-ip. For Kafka-based consumer, we can fill in all 3 fields. For ZK-based consumer, we can just fill in member-id. We can leave the client-id and client-ip part empty since that are not stored explicitly. 

```
GROUP                          TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             OWNER                                         MEMBER-ID
test-consumer-group            test                           0          2               2               0               test-consumer-group_Juns-MacBook-Pro.local-1466121481202-cb9a4849-0 test-consumer-group_Juns-MacBook-Pro.local-1466121481202-cb9a4849
test-consumer-group            test                                                                                                                                    test-consumer-gr
```

vahidhashemian replied:
That's fair. I'll update the code to return the output in the format you suggested. I'm thinking of always returning an empty string for client id and ip for ZK-based consumer (as you suggested), and returning "none" for member id when no active consumer exists for the group.

junrao commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -156,47 +163,102 @@ object ConsumerGroupCommand {
+      var groupMemberIds = zkUtils.getConsumersInGroup(group)
+
+      // mapping of ephemeral owner -> owner
+      val ownerByEOwner = scala.collection.mutable.Map[Long, String]()
```
Could all those scala.collection.mutable.Map be just mutable.Map?

vahidhashemian replied:
I'll make this change too in the next update.

junrao commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -90,37 +90,44 @@ object ConsumerGroupCommand {
 
     protected def describeGroup(group: String): Unit
 
     protected def describeTopicPartition(group: String,
```
Could we rename this to describeMemberTopicPartitions() to make it clear?

vahidhashemian replied:
Sure. Will do in the next update.

junrao commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -90,37 +90,44 @@ object ConsumerGroupCommand {
+                                         getOwner: TopicAndPartition => Option[String],
+                                         getMemberId: TopicAndPartition => Option[String]): Unit = {
+      if (topicPartitions.isEmpty)
+        describePartition(group, Some(topic), None, None, getOwner(null), getMemberId(null))
```
In this case, since the member has no associated partitions, do we need to pass in topic at all? Could we just pass in None?

vahidhashemian replied:
Yes, None should work too. Will update the PR.

vahidhashemian commented:
@junrao Thank you for your feedback. The PR was updated to address your comments.

junrao commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -90,37 +90,46 @@ object ConsumerGroupCommand {
 
-    protected def describeTopicPartition(group: String,
+    protected def describeMemberTopicPartition(group: String,
+                                         topic: String,
```
It doesn't seem that topic is being used. Also, could we fix the indentation?

vahidhashemian replied:
You're right. I'll remove the `topic` parameter and fix indentation.

junrao commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -90,37 +90,46 @@ object ConsumerGroupCommand {
-        }
+                                         getMemberId: TopicAndPartition => Option[String],
+                                         getOwner: TopicAndPartition => Option[String],
+                                         getOwnerHost: TopicAndPartition => Option[String]): Unit = {
```
Could we change getOwner and getOwnerHost to getClientId and getMemberHost accordingly? Also, I am not sure why we getMemberId, getOwner and getOwnerHost need to be a function. Could we just pass in an Option and get rid of "get"?

vahidhashemian replied:
Sure, I'll update in the next patch.

junrao commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -90,37 +90,46 @@ object ConsumerGroupCommand {
 
     protected def printDescribeHeader() {
-      println("%-30s %-30s %-10s %-15s %-15s %-15s %s".format("GROUP", "TOPIC", "PARTITION", "CURRENT-OFFSET", "LOG-END-OFFSET", "LAG", "OWNER"))
+      println("%-30s %-30s %-10s %-15s %-15s %-10s %-50s %-30s %s".format("GROUP", "TOPIC", "PARTITION", "CURRENT-OFFSET", "LOG-END-OFFSET", "LAG", "MEMBER-ID", "OWNER", "OWNER-HOST"))
```
Could we change owner to client-id? Could owner-host be member-host since not every member owns a partition?

vahidhashemian replied:
Sure.

junrao commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -90,37 +90,46 @@ object ConsumerGroupCommand {
-                                  ownerOpt: Option[String]) {
+                                  memberOpt: Option[String],
+                                  ownerOpt: Option[String],
+                                  ownerHostOpt: Option[String]) {
```
Could we change ownerOpt and ownerHostOpt to clientIdOpt and memberHostOpt, respectively?

vahidhashemian replied:
Yup, and I'm going to switch the last two parameters so member related parameters are next to each other. I hope that's fine.

junrao commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -90,37 +90,46 @@ object ConsumerGroupCommand {
       }
       getLogEndOffset(topic, partition) match {
         case LogEndOffsetResult.LogEndOffset(logEndOffset) => print(Some(logEndOffset))
+        case LogEndOffsetResult.NoTopicPartition => print(offsetOpt)
```
If topicPartition doesn't exist, should we really pass in offsetOpt? It seems it's more intuitive to pass in None.

vahidhashemian replied:
Right, will change to `None`.

junrao commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -156,47 +165,96 @@ object ConsumerGroupCommand {
+        val owner = data._1
+        if (!owner.isEmpty) {
+          val eOwner = data._2.getEphemeralOwner()
+          eOwnerByTopicPartition.put(topicPartition, eOwner)
```
Not sure why we need to use ephemeral owner here. It seems that topicPartition->owner gives us the mapping from topic partition to group member id directly. consumerGroupDir + "/ids gives us all members. From these two, we know members that don't own any partitions.

hachikuji replied:
It seems the node of the partition owner includes the threadId. The pattern is always "{consumerId}-{threadId}", so checking the prefix of the partition owner would always get us the right id. Maybe it's a little nicer to use that than the ephemeral owner?

vahidhashemian replied:
Sure I'll try to use that (by ignoring the threadId part) instead of ephemeral owner.

junrao commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -156,47 +165,96 @@ object ConsumerGroupCommand {
+      groupMemberIds.sortBy { - memberTopicPartitions.get(_).size }.foreach { memberId =>
+        topicsByMemberId(memberId).foreach { topic =>
+          val topicPartitions = memberTopicPartitions.getOrElse(memberId, Array()).filter { _.topic == topic }
+          val partitionOffsets = getPartitionOffsets(group, topicPartitions, channelSocketTimeoutMs, channelRetryBackoffMs)
```
This is the case for members not owning any topic partitions, right? If so, could we just set topicPartitions and partitionOffsets to empty?

vahidhashemian replied:
That's correct and makes sense. Will do.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -90,37 +90,46 @@ object ConsumerGroupCommand {
     protected def opts: ConsumerGroupCommandOptions
 
-    protected def getLogEndOffset(topic: String, partition: Int): LogEndOffsetResult
+    protected def getLogEndOffset(topic: Option[String], partition: Option[Int]): LogEndOffsetResult
```
It seems weird to let this function accept `None` for either of these fields (why would I ever try to get the LEO if I don't have a topic or a partition?). Maybe the function can accept an instance of `TopicAndPartition` and the caller can make sure they have an instance prior to calling?

vahidhashemian replied:
I'll try to fix this as suggested.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -90,37 +90,46 @@ object ConsumerGroupCommand {
-                                  topic: String,
-                                  partition: Int,
+                                  topic: Option[String],
+                                  partition: Option[Int],
```
Basically the same comment as above: it's weird to have a function `describePartition` where the partition is optional. Could we take the logic for handling that case out of this function?

vahidhashemian replied:
I understand the concern. Will make some changes to fix the issue.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -156,47 +165,96 @@ object ConsumerGroupCommand {
       val channelSocketTimeoutMs = props.getProperty("channelSocketTimeoutMs", "600").toInt
       val channelRetryBackoffMs = props.getProperty("channelRetryBackoffMsOpt", "300").toInt
+      if (!zkUtils.getConsumerGroups().contains(group)) {
+        println("The consumer group provided does not exist")
```
Minor: would it be helpful to echo the group back to the user in quotes (so that formatting errors are apparent)?

vahidhashemian replied:
That's a good suggestion. Will update this one and the one a few lines below to return the group name as part of the output message.

hachikuji commented:
@vahidhashemian I'm wondering about approaches for testing this patch. One possibility is to parse the output in the test case and build assertions off from that. Another option might be to decouple the command logic from the print logic so that you can build test cases using a convenient object representation. What do you think?

vahidhashemian commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -156,47 +165,96 @@ object ConsumerGroupCommand {
+        val owner = data._1
+        if (!owner.isEmpty) {
+          val eOwner = data._2.getEphemeralOwner()
+          eOwnerByTopicPartition.put(topicPartition, eOwner)
```
I think I had tried that combination before to extract the info. The issue is when I try `get /consumers/group1/owners/test/0` the output looks like this: `cgroup1_kafka-1466461259713-759adaaa-0`. However, when I try `ls /consumers/group1/ids` the output is like `[cgroup1_kafka-1466462141465-ff19e1a5, cgroup1_kafka-1466461259713-759adaaa]`.

It seems the first call returns a client id, and the second one a list of member ids, and that's why the outputs do not quite match (`cgroup1_kafka-1466461259713-759adaaa-0 != cgroup1_kafka-1466461259713-759adaaa`). That's why I did not use this approach as I wasn't sure if there is anything further I can assume to connect the two outputs. If there is, please let me know.

vahidhashemian replied:
@junrao I was wondering what your opinion is about the response above to your comment. This PR has gone through a few more rounds of reviews thanks to @hachikuji and this currently is the only outstanding item. Thanks in advance for looking into this.

vahidhashemian commented:
@junrao @hachikuji Thank you very much for your careful review and feedback. I made some changes in the new patch based on the feedback received.

@junrao The only part that is outstanding in addressing your feedback is [your comment about how to find members with no topic partition](https://github.com/apache/kafka/pull/1336#discussion_r67726141). I responded to your comment and will wait for your clarification as I may be missing something.

@hachikuji Both approaches you suggested for designing unit tests for this patch should work. Parsing the output and writing assertions based on that would be faster but probably not the cleaner option, in my opinion. Separating the logic from reporting seems more reasonable to me but it would require more time and thinking. I can take either direction once I hear your (and others') feedback about it.

hachikuji commented:
@vahidhashemian Yeah, I agree that separating the logic would make testing easier. It's more painful now, but making the code more testable would probably pay off in the long run. I wonder if we could introduce a simple trait to encapsulate the output writing. For example:

``` scala
trait OutputWriter {
  def printError(msg)
  def printAssignmentHeader()
  def printAssignment(topic, partition, memberId,.... etc)
}
```

Then we could test the command using a mock implementation. Anyway, if the level of effort to get some basic coverage using this or another pattern is not too high, I'd suggest we do it here. Otherwise, we could create a followup JIRA.

vahidhashemian commented:
@hachikuji Thanks for your suggestion on how to go about separating the logic. I'll give it a try and let you know how it goes.

vahidhashemian commented:
@hachikuji I've spent some time on this and it seems to involve restructuring of the code to some extent (assuming I haven't over-complicated things). So I'm wondering whether we should keep the changes simpler and do this work and unit tests in separate work items.

hachikuji commented:
@vahidhashemian Either way works for me. I don't think there's any particular hurry to get this patch in since it will probably go in 0.10.1, so if restructuring the code seems like a good idea, I'm probably slightly in favor of doing it here so we don't forget about it later. That said, I know it can be painful to fill in some of these testing gaps (they're gaps because testing wasn't easy!), so if @junrao is OK with it, then doing it separately seems fine to me.

vahidhashemian commented:
@hachikuji Sounds good. I'm okay both ways and will wait for @junrao's feedback.

Regarding the unit tests, I'm trying to wrap my head around how to write them with proper mocking. You can see [here](https://github.com/vahidhashemian/kafka/compare/KAFKA-3144...vahidhashemian:KAFKA-3144-WITH-UNIT-TESTS?expand=1) how I've made changes to the last patch so the `OutputWriter` trait that you suggested is used. The issue now is I don't see clearly how to write the unit tests with mocking (for example, for `--describe` the main functionality is inside the protected `describeGroup()` method). This tells me perhaps further restructuring is required to make the command properly mockable, but I wanted to ask your opinion before heading down that road. Thanks.

vahidhashemian commented:
@hachikuji I restructured `ConsumerGroupCommand.scala` a little bit for testing purposes according to your suggestion and added a number of unit tests. I would appreciate your feedback.

vahidhashemian commented:
@hachikuji This in another pending PR that conflicts with changes required for [KAFKA-3853](https://issues.apache.org/jira/browse/KAFKA-3853), and would be great if we can finalize it soon. Thanks.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -71,64 +72,129 @@ object ConsumerGroupCommand {
-        println("Error while executing consumer group command " + e.getMessage)
-        println(Utils.stackTrace(e))
+        consumerGroupOutputWriter.printError("Executing consumer group command failed due to " + e.getMessage
+            + "\n" + Utils.stackTrace(e))
```
minor: maybe we could just debug log the stack trace?

vahidhashemian replied:
I believe this is being handled in [this other PR](https://github.com/apache/kafka/pull/1548). If it's okay I would let it come through that PR. One of the two PRs would have to be rebased depending on which one goes in first.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -71,64 +72,129 @@ object ConsumerGroupCommand {
     }
   }
 
+  case class ConsumerGroupAssignment(group: String, topicOpt: Option[String], partitionOpt: Option[Int], offsetOpt: Option[Long],
```
nit: maybe something like `PartitionAssignmentState` would be more accurate?

hachikuji replied:
another small nit: the "Opt" suffix is kind of annoying. Could we drop it?

vahidhashemian replied:
Both sound good. Will update in the next commit.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -71,64 +72,129 @@ object ConsumerGroupCommand {
+                                     clientIdOpt: Option[String], logEndOffset: Option[Long])
+
+  trait OutputWriter {
+    def printError(msg: String)
```
Maybe we could change this to something like this:

``` scala
def handleError(e: Throwable)
```

That might make testing easier.

hachikuji replied:
Or maybe we keep the current name and just change the argument type since this is `OutputWriter`.

hachikuji replied:
Actually I see that we print some context-specific error messages, so maybe we need to allow the message to come through. But perhaps we could pass the exception in an optional argument so that we can have a common place to log the stacktrace.

vahidhashemian replied:
Sure, I'll make necessary changes for this.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -71,64 +72,129 @@ object ConsumerGroupCommand {
+  trait OutputWriter {
+    def printError(msg: String)
+    def printConfirmation(msg: String)
+    def printAssignmentHeader()
```
I'm wondering if we can collapse these bottom 3 methods into a single `printAssignment(assignment: Array[ConsumerGroupAssignment])`. Doesn't seem like we're saving that much with the abstract for loop.

vahidhashemian replied:
That's fair. Would it make sense to use `Assignments`/`assignments` to imply all assignments, and `Assignment`/`assignment` to imply a single assignment row?

hachikuji replied:
Not sure I understand the question. I was thinking that a single `printAssignment` method could accept the complete assignment for the full group. Then printing assignment rows or whatever is just an implementation detail. Does that make sense or not?

vahidhashemian replied:
I agree with merging those three methods. I was just curious about naming convention for the method and the variables used to implement them. Maybe [this line](https://github.com/apache/kafka/pull/1336/files#diff-f1b330624af520e6f9824bd308e21f24R105) helps, where I use `assignments` as the full list of assignments, and then each individual member is an `assignment`. I thought I could use this convention across the board. This is very minor and not a big deal though if it's confusing.

hachikuji replied:
Ah I see. maybe we could use `groupAssignment` and `memberAssignment`? As long as it's clear in its context, either way seems ok to me.

vahidhashemian replied:
Yeah, that works too, and would probably be more descriptive. Thanks.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -71,64 +72,129 @@ object ConsumerGroupCommand {
-    def describe() {
-      describeGroup(opts.options.valueOf(opts.groupOpt))
+    def describe(): Option[Array[ConsumerGroupAssignment]] = {
+      describeGroup(opts.options.valueOf(opts.groupOpt)) match {
```
nit: the name `describeGroup` no longer seems quite right. Maybe `assignmentState` would be more accurate since that's what the method returns.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -71,64 +72,129 @@ object ConsumerGroupCommand {
+
+    protected def describeGroup(group: String): Option[Array[ConsumerGroupAssignment]]
+
+    protected def describeMemberTopicPartition(group: String,
```
Maybe `describeGroup` could be `collectGroupAssignment` and this method could be `collectMemberAssignment`?

vahidhashemian replied:
Sure. I was wondering if we should use a `get` prefix (e.g. `getGroupAssignment`) to indicate there is some return value. But I don't see that respected as a convention everywhere in the code.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -152,51 +218,91 @@ object ConsumerGroupCommand {
-      val topicPartitionMap = zkUtils.getPartitionsForTopics(Seq(topic))
-      val partitions = topicPartitionMap.getOrElse(topic, Seq.empty)
-      partitions.map(TopicAndPartition(topic, _))
+    private def getTopicsPartitions(topics: Seq[String]): Seq[TopicAndPartition] = {
```
suggestion: `getAllTopicPartitions`?

vahidhashemian replied:
Sounds good.

hachikuji commented on core/src/test/scala/unit/kafka/admin/ConsumerGroupCommandTest.scala:
```diff
@@ -0,0 +1,140 @@
+import kafka.utils.TestUtils
+
+
+class ConsumerGroupOutputWriterTest extends OutputWriter {
```
Is this used anywhere?

vahidhashemian replied:
No. Thanks for catching this. Will be removed.

vahidhashemian commented:
@hachikuji  Thanks for the thorough feedback. I hope I addressed all of them in the recent update.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -71,64 +72,122 @@ object ConsumerGroupCommand {
+    def printError(msg: String, e: Option[Throwable] = None): Unit = {
+      println("Error: " + msg)
+      if (e.isDefined)
+        debug(Utils.stackTrace(e.get))
```
Minor: if we use debug(), do we need to extract the stack trace? Maybe you could do something like this:

``` scala
e.map(debug("Exception in consumer group command", _))
```

vahidhashemian replied:
Sure, I'll make this change.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -71,64 +72,122 @@ object ConsumerGroupCommand {
 
-    def describe() {
-      describeGroup(opts.options.valueOf(opts.groupOpt))
+    def describe(): Option[Array[PartitionAssignmentState]] = {
```
It's a little weird to have describe() return the partition state. In the test case below using the mock of `OutputWriter`, could you replace the assertions on the result of describe() with `EasyMock.expect()` assertions?

vahidhashemian replied:
This is what I originally wanted to do, but struggled with how to verify the actual result (what would be printed as a result of `describe()` call) against how I call `EasyMock.expect(...)`, which is something like `EasyMock.expect(outputWriterMock.printError("The consumer group 'missing.group' does not exist"))` for the first unit test (`testDescribeNonExistingGroup`). How do I make sure that a call to `consumerGroupCommand.describe()` would actually make the call identified in `EasyMock.expect`?

hachikuji replied:
That's kind of unfortunate. So here's another idea (feel free to dismiss it if it doesn't make sense). Maybe having `describe()` return the assignment is actually heading in the right direction. What if we get rid of `OutputWriter` and move the printing logic into the main method. Then the `ConsumerGroupService` becomes more functional (and testable). For `list()`, you can have it return the list of groups instead of printing them. The trickier one seems to be `delete()`. Maybe you can leave it as it is.

vahidhashemian replied:
Sure, I'll try that. Having `describe()` return the assignment would make testing much easier.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -71,64 +72,122 @@ object ConsumerGroupCommand {
 
-    protected def printDescribeHeader() {
-      println("%-30s %-30s %-10s %-15s %-15s %-15s %s".format("GROUP", "TOPIC", "PARTITION", "CURRENT-OFFSET", "LOG-END-OFFSET", "LAG", "OWNER"))
+    protected def getDescribeResult(group: String,
```
Seems like the only thing this method is contributing is the computation of lag. Maybe we could replace it with a method like this:

``` scala
protected def getLag(offset: Option[Long], logEndOffset: Option[Long])
```

What do you think?

vahidhashemian replied:
Yup, that makes sense.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -71,64 +72,122 @@ object ConsumerGroupCommand {
+
+  trait OutputWriter {
+    def printError(msg: String, e: Option[Throwable] = None)
+    def printConfirmation(msg: String)
```
Based on the usage below, would `printMessage` be a more accurate name?

vahidhashemian replied:
Yes, it would be more appropriate. I'll change it. Thanks.

hachikuji commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -775,6 +775,17 @@ class ZkUtils(val zkClient: ZkClient,
     consumersPerTopicMap
   }
 
+  def getTopicsPerMemberId(group: String, excludeInternalTopics: Boolean) : mutable.Map[String, List[String]] = {
```
Would it make sense to default `excludeInternalTopics` to true?

vahidhashemian replied:
I guess it would. There is an existing method before this one (`getConsumersPerTopic`) that also takes `excludeInternalTopics` and assumes no default value for it. I wanted to make the new method consistent with that one.

vahidhashemian commented:
@hachikuji Thanks for another round of reviews. The PR is updated to address your comments.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -71,64 +78,99 @@ object ConsumerGroupCommand {
+  sealed trait ConsumerGroupService {
 
-    protected def describeGroup(group: String): Unit
+    def list(): List[String]
```
This is looking promising. Maybe we could rename `list` to `listGroups` and `describe` to `describeAssignment`?

vahidhashemian replied:
Sounds good. Wouldn't `describeGroup` (singular) be more self-explanatory than `describeAssignment`?

hachikuji replied:
Yeah, that sounds good to me.

vahidhashemian replied:
And we could also rename `delete` to `deleteGroups`?

hachikuji replied:
Makes sense to me.

vahidhashemian commented:
Thanks @hachikuji. Made those changes.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -283,50 +363,55 @@ object ConsumerGroupCommand {
       adminClient.describeConsumerGroup(group) match {
-        case None => println(s"Consumer group `${group}` does not exist.")
+        case None =>
+          consumerGroupOutputWriter.printError(s"Consumer group '$group' does not exist.")
```
Looks like we would print this twice: once here and once in main. I'm wondering if we could remove `ConsumerGroupOutputWriter` from this class and do all the output in main. Maybe we just need to move the empty check that you have below in main?

vahidhashemian replied:
Thanks for catching this. I'll fix in the next update. And we can remove `ConsumerGroupOutputWriter` from that class, as you suggested.

vahidhashemian committed:
This PR makes a few enhancements to the --describe option of ConsumerGroupCommand:
1. Listing members with no assigned partitions.
2. Reporting the member id along with the owner of each partition (owner is supposed to be the logical application id and all members in the same group are supposed to set the same owner).
3. Printing a message indicating whether ZooKeeper based or new consumer API based information is being reported.

It also adds unit tests to verify the added functionality.

Note: The third request on the corresponding JIRA (listing active offsets for empty groups of new consumers) is not implemented as part of this PR, and has been moved to its own JIRA [KAFKA-3853](https://issues.apache.org/jira/browse/KAFKA-3853).

vahidhashemian commented:
@junrao The patch is updated to include coordinator info when `--new-consumer` is used. Since the group column and the new coordinator column report redundant information I modified the output of `--new-consumer` option to

```
Note: This will only show information about consumers that use the new consumer API (non-ZooKeeper-based consumers).

Consumer group: cgroup1
Consumer group coordinator broker id: 0

TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG        MEMBER-ID                                          MEMBER-HOST                    CLIENT-ID
bar                            0          234             234             0          consumer-1-a0d7940a-c90a-40df-b995-64d2e104bbb6    /127.0.0.1                     consumer-1
```

The ZooKeeper based output will not print the coordinator line above.

Your feedback is appreciated.

hachikuji commented on core/src/main/scala/kafka/coordinator/GroupMetadata.scala:
```diff
@@ -128,6 +129,15 @@ case class GroupSummary(state: String,
+ * Case class used to represent group metadata (including the group coordinator) for the DescribeGroup API
+ */
+case class ConsumerGroupSummary(state: String,
+                                protocolType: String,
```
Is this needed for a consumer group? I think protocol type will always be "consumer."

vahidhashemian replied:
Yes, it'll be `consumer`, but there is a check [here](https://github.com/vahidhashemian/kafka/blob/904bcee920605e1fe5417fe39b5397194b665832/core/src/main/scala/kafka/admin/AdminClient.scala#L151) to verify that the protocol type os valid. If we remove this, I think that check has to be removed too. Should I still go ahead and remove it?

hachikuji replied:
I would probably move that check into `describeGroup`. Now that I'm thinking about it, we should probably either rename `describeGroup` to `describeConsumerGroup`, or we should let `describeGroup` return a generic `GroupSummary` while `describeConsumerGroup` returns `ConsumerGroupSummary`.

vahidhashemian replied:
Thanks. With the first suggestion, there already is a `describeConsumerGroup` method. Unless you prefer this suggestion (using a different method name) I'll go ahead with your second suggestion.

vahidhashemian replied:
Upon further consideration, I think I'm going to adopt your first suggestion, because I need the current `describeGroup` method to return the coordinator as part of its output; so I can report the coordinator's broker id. I'm going to use the method name `getConsumerGroupSummary` instead.

hachikuji commented on core/src/main/scala/kafka/coordinator/GroupMetadata.scala:
```diff
@@ -128,6 +129,15 @@ case class GroupSummary(state: String,
+ */
+case class ConsumerGroupSummary(state: String,
+                                protocolType: String,
+                                protocol: String,
```
For consumer groups, the protocol is really the assignment strategy.

vahidhashemian replied:
Sure, I'll rename this field to `assignmentStrategy`.

hachikuji commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -802,6 +802,17 @@ class ZkUtils(val zkClient: ZkClient,
+  def getTopicsPerMemberId(group: String, excludeInternalTopics: Boolean = true) : mutable.Map[String, List[String]] = {
+    val dirs = new ZKGroupDirs(group)
+    val memberIds = getChildrenParentMayNotExist(dirs.consumerRegistryDir)
+    val topicsPerMemberIdMap = new mutable.HashMap[String, List[String]]
```
As far as I can tell, the return type of this method doesn't need to be mutable. Maybe something like this would be a little nicer?

``` scala
    memberIds.map { memberId =>
      val topicCount = TopicCount.constructTopicCount(group, memberId, this, excludeInternalTopics)
      memberId -> topicCount.getTopicCountMap.keys.toList
    }.toMap
```

(Note I used `keys` instead of `keySet` since we throw away the set anyway.)

vahidhashemian replied:
Makes sense. Thank you for the suggestion. I'll update the method.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -300,36 +394,39 @@ object ConsumerGroupCommand {
     }
 
-    protected def getLogEndOffset(topic: String, partition: Int): LogEndOffsetResult = {
+    protected def getLogEndOffset(topicAndPartition: TopicAndPartition): LogEndOffsetResult = {
```
Would it make a big difference if this method accepted `TopicPartition` instead of `TopicAndPartition` since it seems that's what we need anyway?

vahidhashemian replied:
I'll give it a try in the next patch.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -379,9 +476,9 @@ object ConsumerGroupCommand {
 
   class ConsumerGroupCommandOptions(args: Array[String]) {
-    val ZkConnectDoc = "REQUIRED (only when using old consumer): The connection string for the zookeeper connection in the form host:port. " +
+    val ZkConnectDoc = "REQUIRED (unless new-consumer is used): The connection string for the zookeeper connection in the form host:port. " +
```
Is this intentional? Same for the couple doc changes below.

vahidhashemian replied:
To be honest, it's been so long that I don't recall why I made these changes. But when I try the current command, I can't tell what the meaning of "new consumer being the default" is. It seems to me that we need to either provide `--zookeeper` or `--bootstrap-server`; it's not like if we use `--new-consumer` we don't have to provide `--bootstrap-server`. The only restriction around `--new-consumer` seems to be that it cannot be used along with `--zookeeper`. The descriptions in parenthesis are not very clear to me. But it might be just me. Your thoughts?

hachikuji replied:
I agree it's not super clear. Maybe we should just avoid saying it's required since it saves us from needing to qualify? You can try to fix this if you want, but I'd be ok just leaving it as it is since it seems orthogonal to the rest of this.

vahidhashemian replied:
Sure, I'll revert these changes. I'm OK with leaving the "required" text in since other tools have it too, but we need to better qualify them, as you mentioned. I may work on it separately later.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,83 +57,137 @@ object ConsumerGroupCommand {
+                                      memberHost: Option[String], clientId: Option[String], logEndOffset: Option[Long])
 
-    def list(): Unit
+  class ConsumerGroupOutputWriter extends Logging {
```
Since we didn't end up needing this for testing, maybe we can just get rid of it and keep its methods one level up?

vahidhashemian replied:
Sure, sounds good.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,83 +57,137 @@ object ConsumerGroupCommand {
-      describeGroup(opts.options.valueOf(opts.groupOpt))
+    def printError(msg: String, e: Option[Throwable] = None): Unit = {
+      println("Error: " + msg)
+      if (e.isDefined)
```
nit: Since you're using `map`, we don't need this check.

vahidhashemian replied:
Thanks for catching this.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,83 +57,137 @@ object ConsumerGroupCommand {
+    def printMessage(msg: String): Unit = println(msg)
 
-    protected def opts: ConsumerGroupCommandOptions
+    def printAssignment(groupAssignment: Array[PartitionAssignmentState]): Unit = {
```
Could we let this function handle the empty assignment case as well?

vahidhashemian replied:
We could, but we would need to pass more arguments now that we want to distinguish between empty group and rebalancing group. Plus, in the case of an empty assignment we are using the `printError` method, instead of actually printing assignments. Do you still think we should change it?

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,83 +57,137 @@ object ConsumerGroupCommand {
+            consumerGroupOutputWriter.printError(s"The consumer group '${opts.options.valuesOf(opts.groupOpt).asScala.head}' does not exist.")
+          case Some(assignments) =>
+            if (assignments.isEmpty)
+              consumerGroupOutputWriter.printError(s"Consumer group '${opts.options.valuesOf(opts.groupOpt).asScala.head}' is rebalancing.")
```
I'm not actually sure this message is still correct. In KAFKA-2720, we introduced an `Empty` state for the group, which basically persists until all the offsets for the group have expired. In this case, the assignment will be empty, but it will not be rebalancing.

vahidhashemian replied:
You're right. Would it be OK to change this message to `Consumer group ... has no active member or is rebalancing`?

hachikuji replied:
It would be more ideal if we could tell the user which is the case. Maybe we need to propagate the state of the group down to this method.

vahidhashemian replied:
Yup, will try to do this in the next patch.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,83 +57,137 @@ object ConsumerGroupCommand {
   }
 
-  sealed trait ConsumerGroupService {
+  case class PartitionAssignmentState(group: String, coordinator: Option[Node], topic: Option[String], partition: Option[Int],
```
Should this be `private`?

vahidhashemian replied:
This `case class` is used in other classes in the same file. Since it's defined outside the scope of those classes `private` wouldn't work, but `protected` would. I hope I didn't misunderstood your point.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -155,51 +211,91 @@ object ConsumerGroupCommand {
 
-    protected def getLogEndOffset(topic: String, partition: Int): LogEndOffsetResult = {
-      zkUtils.getLeaderForPartition(topic, partition) match {
+    protected def getLogEndOffset(topicAndPartition: TopicAndPartition): LogEndOffsetResult = {
```
Similar to other comment: maybe we could use `TopicPartition` here.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -155,51 +211,91 @@ object ConsumerGroupCommand {
+    private def getAllTopicPartitions(topics: Seq[String]): Seq[TopicAndPartition] = {
+      val topicPartitionMap = zkUtils.getPartitionsForTopics(topics)
+      var topicPartitions: Seq[TopicAndPartition] = Seq.empty
+      topics.foreach { topic =>
```
Why not use `map`?

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -155,51 +211,91 @@ object ConsumerGroupCommand {
+      var groupMemberIds = zkUtils.getConsumersInGroup(group)
+
+      // mapping of topic partition -> ephemeral owner
+      val eOwnerByTopicPartition = mutable.Map[TopicAndPartition, Long]()
```
Seems like this is another case where we don't actually need a mutable collection if we use `flatMap`.

vahidhashemian replied:
I'll try to make use of `flatMap`.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -155,51 +211,91 @@ object ConsumerGroupCommand {
+      topicPartitions.foreach { topicPartition =>
+        val data = zkUtils.readDataMaybeNull(new ZKGroupTopicDirs(group, topicPartition.topic).consumerOwnerDir + "/" + topicPartition.partition)
+        val owner = data._1
+        if (!owner.isEmpty) {
```
Maybe change this to `map`, then I think `flatMap` as suggested above would work nicely.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -155,51 +211,91 @@ object ConsumerGroupCommand {
+      // mapping of ephemeral owner -> group member id
+      val memberIdByEOwner = groupMemberIds.flatMap { memberId =>
+        val eOwner = zkUtils.readDataMaybeNull(new ZKGroupDirs(group).consumerGroupDir + "/ids/" + memberId)._2.getEphemeralOwner()
+        Some(eOwner).map { eOwner => eOwner -> memberId }
```
This looks odd. We create an option just so we can call `map`? Could we replace this with:

``` scala
eOwner -> memberId
```

Same for the loop below.

vahidhashemian replied:
This section will be removed now that we decided to use the owner info directly.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -155,51 +211,91 @@ object ConsumerGroupCommand {
+        val memberId = memberIdByTopicPartition.get(topicPartition)
+        // since member id is repeated in client id, leave member host and client id empty
+        assignmentRows = assignmentRows ++ collectMemberAssignment(group, None, List(topicPartition), partitionOffsets.get, memberId, None, None)
+        if (!memberId.isEmpty) {
```
Using `map` would be nicer?

hachikuji reviewed (commented):
Sorry for leaving this hanging for so long. Did another pass and have some mostly minor suggestions.

vahidhashemian commented:
@hachikuji Thanks for the thorough feedback. I hope I was able to address your comments in the recent patch.

hachikuji commented on core/src/main/scala/kafka/admin/AdminClient.scala:
```diff
@@ -121,44 +121,44 @@ class AdminClient(val time: Time,
-    } else {
-      Some(List.empty)
-    }
+  def describeConsumerGroup(groupId: String): (String, Option[(Node, List[ConsumerSummary])]) = {
```
If we had `ConsumerGroupSummary` include a list of `ConsumerSummary` objects instead of `MemberSummary`, would we still need this function?

vahidhashemian replied:
I'm not sure if I follow. Could you please elaborate a bit?

hachikuji replied:
Currently `ConsumerGroupSummary` has a field for the group members, which are represented as instances of `MemberSummary`. I'm wondering if it would make sense to use `ConsumerSummary` instead. Then we might only need a single method returning `ConsumerGroupSummary`.

vahidhashemian replied:
Do you mean merging `getConsumerGroupSummary` and `describeConsumerGroup` into one method that returns a `ConsumerGroupSummary` object? If so, I thought about it when I was making the recent changes, and noticed that `getConsumerGroupSummary` is being used in a few other places. That's why I hesitated to make the unnecessary change. But I guess they could be merged and all calls to `getConsumerGroupSummary` would become calls to `describeConsumerGroup`. Please correct me if I'm misunderstood. Thanks.

hachikuji replied:
Yeah, that's what I meant. Don't bother if it's a ton of additional work, but seems like a nice cleanup.

vahidhashemian replied:
No problem, I'll give it try in the next patch.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +56,132 @@ object ConsumerGroupCommand {
+    e.map(debug("Exception in consumer group command", _))
+  }
+
+  def printMessage(msg: String): Unit = println(msg)
```
Seems like this method isn't giving us much anymore. Maybe it's ok to use `println` directly?

hachikuji commented on core/src/main/scala/kafka/coordinator/GroupMetadata.scala:
```diff
@@ -128,6 +129,14 @@ case class GroupSummary(state: String,
 /**
+ * Case class used to represent group metadata (including the group coordinator) for the DescribeGroup API
+ */
+case class ConsumerGroupSummary(state: String,
```
It's a little weird to locate this in `kafka.coordinator` since the coordinator is technically agnostic to group internals. The best alternative I can think of is maybe to put it in tools with `AdminClient`. What do you think?

vahidhashemian replied:
Sure, it makes sense. I'll move it.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -300,36 +377,41 @@ object ConsumerGroupCommand {
-    def list() {
-      adminClient.listAllConsumerGroupsFlattened().foreach(x => println(x.groupId))
+    def listGroups(): List[String] = {
+      adminClient.listAllConsumerGroupsFlattened().map { _.groupId }.toList
```
nitpick: since it's a simple statement, maybe parenthesis in `map` would be a little nicer than braces?

hachikuji commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -802,6 +802,15 @@ class ZkUtils(val zkClient: ZkClient,
     consumersPerTopicMap
   }
 
+  def getTopicsPerMemberId(group: String, excludeInternalTopics: Boolean = true) : Map[String, List[String]] = {
```
nitpick: we usually don't put a space before the ':'.

vahidhashemian replied:
I'll remove the space. There are a few other occurrences in this file that I'll fix too.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +56,132 @@ object ConsumerGroupCommand {
-        consumerGroupService.list()
-      else if (opts.options.has(opts.describeOpt))
-        consumerGroupService.describe()
+        consumerGroupService.listGroups().foreach { printMessage(_) }
```
nitpick: I think parenthesis are a little nicer for simple one-liners like this.

hachikuji replied:
There are few other places in the patch where we could also change this.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +56,132 @@ object ConsumerGroupCommand {
-        }
+    protected def getLogEndOffset(topicPartition: TopicPartition): LogEndOffsetResult
+
+    protected def collectGroupAssignment(group: String): (String, Option[Seq[PartitionAssignmentState]])
```
Since the zookeeper service has no notion of consumer state, should we return `Option[String]` instead?

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -155,51 +211,74 @@ object ConsumerGroupCommand {
+        val owner = zkUtils.readDataMaybeNull(new ZKGroupTopicDirs(group, topicPartition.topic).consumerOwnerDir + "/" + topicPartition.partition)._1.get
+        // cut the appending thread id
+        val splitPos = owner.lastIndexOf('-')
+        Seq(topicPartition -> owner.substring(0, splitPos))
```
If we use normal `map` instead of `flatMap`, does this need to be a `Seq` anymore?

vahidhashemian replied:
No, it doesn't. Thanks for catching it.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -155,51 +211,74 @@ object ConsumerGroupCommand {
+        collectMemberAssignment(group, None, List(topicPartition), partitionOffsets.get, memberId, None, None)
+      }
+
+      groupMemberIds.sortBy { - memberTopicPartitions.get(_).size }.foreach { memberId =>
```
Could this one be a `flatMap` like the one just above?

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -242,13 +319,13 @@ object ConsumerGroupCommand {
         try {
           if (AdminUtils.deleteConsumerGroupInZK(zkUtils, group))
-            println("Deleted all consumer group information for group %s in zookeeper.".format(group))
+            printMessage("Deleted all consumer group information for group '$group' in zookeeper.")
```
Looks like we're missing the 's' at the start of the string. Another few of these below.

vahidhashemian replied:
Thanks, I noticed there were a few more.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -155,51 +211,74 @@ object ConsumerGroupCommand {
           }.getOrElse(LogEndOffsetResult.Ignore)
         case None =>
-          println(s"No broker for partition ${new TopicPartition(topic, partition)}")
+          printError(s"No broker for partition '${topicPartition}'")
```
nitpick: I think the braces are unnecessary if it is a simple variable.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -300,36 +377,41 @@ object ConsumerGroupCommand {
+            else {
+              val consumer = getConsumer()
+              var assignmentRows = Array[PartitionAssignmentState]()
+              consumerSummaries.foreach { consumerSummary =>
```
Looks like another case where we might be able to change the `foreach` to a `map` with a `toArray` at the end.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +56,132 @@ object ConsumerGroupCommand {
-        consumerGroupService.describe()
+        consumerGroupService.listGroups().foreach { printMessage(_) }
+      else if (opts.options.has(opts.describeOpt)) {
+        val describeGroupResult = consumerGroupService.describeGroup()
```
nitpick: Maybe we could destructure on assignment?

``` scala
val (state, assignments) = consumerGroupService.describeGroup()
```

vahidhashemian replied:
Good suggestion.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +56,132 @@ object ConsumerGroupCommand {
+          case Some(assignments) =>
+            if (assignments.isEmpty)
+              if (describeGroupResult._1 == "Empty")
+                printError(s"Consumer group '${opts.options.valuesOf(opts.groupOpt).asScala.head}' has no active member.")
```
nitpick: should 'member' be plural? Also, a little surprising we don't have a variable for the groupId. Maybe we could add one above to make these messages a little easier to read.

vahidhashemian commented:
@hachikuji I submitted another update taking into account your comments. Thanks again for the thorough reviews and constructive feedback.

hachikuji commented on core/src/main/scala/kafka/admin/AdminClient.scala:
```diff
@@ -121,44 +121,35 @@ class AdminClient(val time: Time,
-      val assignment = Utils.readBytes(member.memberAssignment())
-      MemberSummary(member.memberId(), member.clientId(), member.clientHost(), metadata, assignment)
+      val assignment = ConsumerProtocol.deserializeAssignment(ByteBuffer.wrap(Utils.readBytes(member.memberAssignment())))
+      ConsumerSummary(member.memberId(), member.clientId(), member.clientHost(), assignment.partitions.toList)
```
nit: shouldn't need parenthesis for most of these getters, I think. There are a bunch of these around the patch.

vahidhashemian replied:
I'll try to remove them. I see that their occurrences are beyond what's in this patch.

hachikuji replied:
The general rule is to omit the parenthesis if the function does not mutate any state. No need to catch all such cases, it just stood out a bit on this line.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +56,131 @@ object ConsumerGroupCommand {
+  }
+
+  def printAssignment(groupAssignment: Seq[PartitionAssignmentState]): Unit = {
+    println("Consumer group: " + groupAssignment.head.group)
```
nit: might be nice to add a `require` to validate that the group assignment is not empty.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +56,131 @@ object ConsumerGroupCommand {
+            printError(s"The consumer group '$groupId' does not exist.")
+          case Some(assignments) =>
+            if (assignments.isEmpty)
+              if (state == "Empty")
```
This check needs to be updated since `state` is now an `Optional`.

vahidhashemian replied:
You're right. Thanks. I'll fix this in the next patch.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +56,131 @@ object ConsumerGroupCommand {
 
+  def printError(msg: String, e: Option[Throwable] = None): Unit = {
+    println("Error: " + msg)
+    e.map(debug("Exception in consumer group command", _))
```
This should probably be `foreach` since we don't need any the return type.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +56,131 @@ object ConsumerGroupCommand {
+                                          clientIdOpt: Option[String]): Array[PartitionAssignmentState] = {
+      if (topicPartitions.isEmpty)
+        Array[PartitionAssignmentState](
+            new PartitionAssignmentState(group, coordinator, None, None, None, getLag(None, None), memberIdOpt, memberHostOpt, clientIdOpt, None)
```
nit: no need for `new`.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +56,131 @@ object ConsumerGroupCommand {
+      else {
+        var assignmentRows: Array[PartitionAssignmentState] = Array()
+        topicPartitions
+          .sortBy { case topicPartition => topicPartition.partition }
```
nit: maybe this could be

``` scala
.sortBy(_.partition)
```

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -155,51 +210,74 @@ object ConsumerGroupCommand {
+        val partitionOffsets = getPartitionOffsets(group, List(topicPartition), channelSocketTimeoutMs, channelRetryBackoffMs)
+        val memberId = consumerIdByTopicPartition.get(topicPartition)
+        // since member id is repeated in client id, leave member host and client id empty
+        if (!memberId.isEmpty) {
```
nit: maybe this could be a `foreach`?

vahidhashemian replied:
Not sure if I follow this one?

vahidhashemian replied:
Never mind. I think you mean something like

```
memberId.foreach(id => groupMemberIds.filterNot(_ == id))
```

hachikuji replied:
Yeah, something like that. Kind of hard to tell sometimes when you're trying to get a little too cute, but using `map` and `foreach` seems to generally be preferred over an explicit check for `isDefined` or `isEmpty`.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +56,131 @@ object ConsumerGroupCommand {
+                                  memberHostOpt: Option[String],
+                                  clientIdOpt: Option[String]): PartitionAssignmentState = {
+      def getDescribePartitionResult(logEndOffsetOpt: Option[Long]): PartitionAssignmentState =
+        new PartitionAssignmentState(group, coordinator, Option(topic), Option(partition), offsetOpt,
```
nit: no need for `new`

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -300,36 +376,39 @@ object ConsumerGroupCommand {
-    def list() {
-      adminClient.listAllConsumerGroupsFlattened().foreach(x => println(x.groupId))
+    def listGroups(): List[String] = {
+      adminClient.listAllConsumerGroupsFlattened().map(_.groupId).toList
```
nit: unneeded `toList`.

hachikuji commented on core/src/main/scala/kafka/coordinator/GroupMetadata.scala:
```diff
@@ -17,14 +17,13 @@
-
 import kafka.common.OffsetAndMetadata
+import kafka.utils.nonthreadsafe
+import org.apache.kafka.common.Node
```
nit: unneeded import

hachikuji commented on core/src/main/scala/kafka/admin/AdminClient.scala:
```diff
@@ -121,44 +121,35 @@ class AdminClient(val time: Time,
+   */
+  case class ConsumerGroupSummary(state: String,
+                                  assignmentStrategy: String,
+                                  consumerSummaries: Option[List[ConsumerSummary]],
```
Could this be `consumers`. The `Summaries` suffix seems a tad verbose.

vahidhashemian replied:
Sure, sounds fair.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +56,131 @@ object ConsumerGroupCommand {
       case e: Throwable =>
-        println("Error while executing consumer group command " + e.getMessage)
-        println(Utils.stackTrace(e))
+        printError("Executing consumer group command failed due to " + e.getMessage, Some(e))
```
nit: since we've converted most of these to use string interpolation, maybe we could do the same here? I noticed a couple others.

hachikuji reviewed (commented):
Few more nitpicks. I think it's getting close.

vahidhashemian commented:
@hachikuji I had one more crack at this. If you spot any more potential improvements please let me know. Thanks.

BTW, the junkins error seems to be unrelated to this.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +55,136 @@ object ConsumerGroupCommand {
+            printError(s"The consumer group '$groupId' does not exist.")
+          case Some(assignments) =>
+            if (assignments.isEmpty)
+              state.get match {
```
I may have missed it, but how do we know this `get` is safe? Should we match using the option instead?

vahidhashemian replied:
I believe for the ZooKeeper-based consumers the `describeGroup` returns either `None` for `assignments` or some array (and the array cannot be empty because as soon as a group starts there is an assignment row). Therefore, this line would not be reached in that case. And for Java based consumer groups as far as I can tell `state` always has some value, and it cannot be `None`. So this line would be safe to call.

Having said that I'm ok with checking `state` here instead of `state.get`.

One more question. I tried creating a new topic and starting an old consumer consuming from that topic belonging to a new consumer group. When I tried `describe` with this patch I get this output for a few seconds, and then the error vanishes as the initializations are done. Are we OK with this behavior?

```
Note: This will only show information about consumers that use ZooKeeper (not those using the Java consumer API).

Error: Could not fetch offset from zookeeper for group 'cgroup3' partition '[test2,0]' due to missing offset data in zookeeper.
Consumer group: cgroup3

TOPIC    PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG    MEMBER-ID  ...
test2    0                                          0      cgroup3_kafka-1477072199030-5c817e19   ...
```

vahidhashemian commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +55,139 @@ object ConsumerGroupCommand {
+                case None =>
+                  // the control should never reach here
+                  throw new KafkaException("Expected a valid consumer group state, but none found.")
+              }
```
This is the refactored code for `state` check. Please let me know if you see issues with it. Thanks.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +55,139 @@ object ConsumerGroupCommand {
+  }
+
+  def printAssignment(groupAssignment: Seq[PartitionAssignmentState]): Unit = {
+    require(!groupAssignment.isEmpty)
```
nit: you can use `nonEmpty`.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +55,139 @@ object ConsumerGroupCommand {
+      case Some(node) => println(s"Consumer group coordinator broker id: ${node.id}")
+      case default =>
+    }
+    println("\n%-30s %-10s %-15s %-15s %-10s %-50s %-30s %s".format("TOPIC", "PARTITION", "CURRENT-OFFSET", "LOG-END-OFFSET", "LAG", "MEMBER-ID", "MEMBER-HOST", "CLIENT-ID"))
```
Couple naming suggestions:
1. `MEMBER-ID` -> `CONSUMER-ID`
2. `MEMBER-HOST` -> `HOST`

What do you think? Also, it seems neither of these options are available for the old consumer. Maybe it would reduce the noise if we leave them out of the output in that case? Not sure if there's a clean way to do that though.

vahidhashemian replied:
I can make the name changes. Regarding availability for old consumer, it's actually `MEMBER_HOST` and `CLIENT_ID` (the last two) that will be blank in the output.

I think I can leverage the `node` variable (that I introduced to report the coordinator broker id for new consumer option) and based on that decide whether the last two column should be printed or no. I'll submit an update shortly and you can take a look and let me know what you think. Thanks.

hachikuji reviewed (commented):
Thanks for all the effort and patience! One final question and a tiny nitpick and I think this is looking good.

vahidhashemian commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +55,147 @@ object ConsumerGroupCommand {
+        memberAssignment.topic.getOrElse(""), memberAssignment.partition.getOrElse(""),
+        memberAssignment.offset.getOrElse(""), memberAssignment.logEndOffset.getOrElse(""), memberAssignment.lag.getOrElse(""),
+        memberAssignment.consumerId.getOrElse("none")))
+      if (groupAssignment.head.coordinator.nonEmpty)
```
I don't feel very happy about this repeating check. To me it was either this or totally separate print statements for old and new consumers in the `match` block of line 114 above. What do you think?

hachikuji replied:
It might be a little less annoying if you put the result in a val (e.g. `useNewConsumer`). Maybe you could even pass `opts.useOldConsumer` into this function to avoid the need to check the coordinator.

vahidhashemian replied:
Passing `opts.useOldConsumer` sounds good; but that doesn't avoid the need for this repeating `if` block. Or I misunderstood?

hachikuji replied:
Yeah, you still need it. Seems not too bad to me. You could move it outside the loop, but then you'd need to repeat the loop in both arms of the `if`, which seems worse.

vahidhashemian replied:
Right, I'll keep it as is then. Thanks.

vahidhashemian commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +55,147 @@ object ConsumerGroupCommand {
+        memberAssignment.consumerId.getOrElse("none")))
+      if (groupAssignment.head.coordinator.nonEmpty)
+        print("%-30s %s".format(memberAssignment.host.getOrElse(""), memberAssignment.clientId.getOrElse("")))
+      println()
```
BTW, I'm thinking printing "-" instead of "" when data is not available or does not apply would look better and more readable (especially when there are multiple rows with blank columns). Compare

```
TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG        CONSUMER-ID
test1                          0          0               0               0          consumer-1-7913cdd0-e177-409f-86fa-3fcf92ddd926
                                                                                     consumer-1-b2f82fe9-6c1a-4928-8e6d-ccfab0906788
```

with

```
TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG        CONSUMER-ID
test1                          0          0               0               0          consumer-1-7913cdd0-e177-409f-86fa-3fcf92ddd926
-                              -          -               -               -          consumer-1-b2f82fe9-6c1a-4928-8e6d-ccfab0906788
```

hachikuji replied:
Good idea.

vahidhashemian commented:
Just did some cleanup of variable/method/field names based on the recent `member id` to `consumer id` renaming.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -55,77 +55,146 @@ object ConsumerGroupCommand {
+
+  def printAssignment(groupAssignment: Seq[PartitionAssignmentState], useNewConsumer: Boolean): Unit = {
+    require(groupAssignment.nonEmpty)
+    println("Consumer group: " + groupAssignment.head.group)
```
Really sorry to keep adding comments... Did we print this before? It seems unnecessary given that we require the group to be passed on the command line. One downside to having this line and the one below is that it's a little tougher to parse the output. On the other hand, having a way to get the coordinator seems useful for debugging. I wonder if it would make sense to add that to the `--list` option in a separate patch. What do you think?

vahidhashemian replied:
That's fine with me. I added these two lines (along with [this](https://github.com/vahidhashemian/kafka/blob/edb983f2614d4455b12628a698ea701b469dfab2/core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala#L58) and [this](https://github.com/vahidhashemian/kafka/blob/edb983f2614d4455b12628a698ea701b469dfab2/core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala#L61)) because there was a request on the corresponding JIRA asking for some clarification on the printed output.

I'm fine with removing them and have the `--list` option report the coordinator broker id.
Should I also leave out the other two lines I linked to above?

hachikuji replied:
Looked back at the JIRA. Maybe you could print the warning messages to stderr? I also saw Jun's comment about printing the coordinator. It's a little odd to print it as a column in the table as you suggested (since it will be the same for all members), but that might be a better choice. I'd be ok with either doing that or adding it to the `--list` option separately. The latter might be a little nicer since this already has a lot of columns and I think we should have the coordinator information already when using `--list`.

vahidhashemian replied:
I also think using `--list` to report the coordinator id is better for the same reasons you mentioned. I'll submit another update shortly, and open a JIRA for reporting the coordinator.

hachikuji commented:
LGTM. Really appreciate the effort on this PR. It's a huge contribution when you can not only add functionality, but improve the testability of the code.

vahidhashemian commented:
Thanks @hachikuji for your feedback along the way. Very much appreciated.
